File: 1697-Checking-Existence-of-Edge-Length-Limited-Paths/src/Solution.java
import java.util.Arrays;
import java.util.Comparator;

public class Solution {
    int[] fm;

    public int findHead(int node) {
        int fa = fm[node];
        if (fa != node) fa = findHead(fa);
        fm[node] = fa;
        return fa;
    }

    public void union(int a, int b) {
        fm[findHead(b)] = fm[findHead(a)];
    }
    public boolean[] distanceLimitedPathsExist(int n, int[][] edgeList, int[][] queries) {
        fm = new int[n];
        for (int i = 0; i < n; i++) fm[i] = i;
        boolean[] ret = new boolean[queries.length];
        int index = 0;
        Integer[] h = new Integer[queries.length];
        for (int i = 0; i < h.length; i++) h[i] = i;
        Arrays.sort(h, Comparator.comparingInt(o -> queries[o][2]));
        Arrays.sort(edgeList, Comparator.comparingInt(o -> o[2]));
        for (Integer i : h) {
            while (index < edgeList.length && edgeList[index][2] < queries[i][2]) {
                union(edgeList[index][0], edgeList[index][1]);
                index++;
            }
            ret[i] = findHead(queries[i][0]) == findHead(queries[i][1]);
        }
        return ret;
    }
}

File: 1697-Checking-Existence-of-Edge-Length-Limited-Paths/cpp/solution.cpp
#include "bits/stdc++.h"

using namespace std;

const int N = 100010;

struct Node {
    int a, b, c, d;
    bool operator < (const Node & t) const {
        return c < t.c;
    }
} e[N], q[N];


class Solution {
public:
    vector<int> f;

    int find (int x) {
        if (x != f[x]) f[x] = find(f[x]);
        return f[x];
    }

    bool is_connected(int x, int y) {
        return find(x) == find(y); 
    }

    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& ee, vector<vector<int>>& qq) {
        int m = ee.size(), k = qq.size();
       

        for (int i = 0; i < m; i++) 
            e[i] = {ee[i][0], ee[i][1], ee[i][2]};
        for (int i = 0; i < k; i++) 
            q[i] = {qq[i][0], qq[i][1], qq[i][2], i};
        sort(e, e + m), sort(q, q + k);
        f.resize(n);
        for (int i = 0; i < n; i++) f[i] = i;
        vector<bool> res(k);
        for (int i = 0, j = 0; i < k; i++) {
            while (j < m && e[j].c < q[i].c) {
                int a = e[j].a, b = e[j].b;
                f[find(a)] = find(b);
                j++;
            }
            res[q[i].d] = is_connected(q[i].a, q[i].b);
        }
        return res;
        
    }
};